### PR TITLE
Add initial code for the predictor service

### PR DESCRIPTION
**Pull request by LucasCairns**
## Context

This PR add's the initial code for the service layer logic to determine if the available answers on an assessment meet the conditions of it's associated predictors

## Responses

The predictors are returned alongside the updated episode
```json
{
    "episodeId": 2,
    "episodeUuid": "90f2b674-ae1c-488d-8b85-0251708ef6b6",
    "assessmentUuid": "6f3f2c4a-38ac-49ce-b790-70bc170fe553",
    "reasonForChange": "new episode",
    "created": "2019-11-14T08:11:53.177108",
    "answers": {
        "63099aab-f852-4dd9-9179-16ee2218d0c6": [
            "34"
        ]
    },
    "predictors": [
        {
            "type": "RSR",
            "status": "DETERMINED",
            "score": 1234

        }
    ]
}
```

## Request

The request to the ARN API is currently stubbed pending implementation - the request will look something like this and uses the `AnswersDto` to support different types and answers for multiple choice questions 

```json
{
    "type": "RSR",
    "answers": {
        "age_first_sanction": ["34"]
    }
}
```

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/assessments/services/AssessmentUpdateService.kt [hunk 1]
```diff
@@ -67,7 +68,15 @@ class AssessmentUpdateService(
     assessmentRepository.save(episode.assessment)
     log.info("Saved episode ${episode.episodeUuid} for assessment ${episode.assessment?.assessmentUuid}")
 
-    return AssessmentEpisodeDto.from(episode, oasysResult)
+    val updatedEpisode = AssessmentEpisodeDto.from(episode, oasysResult)
```
Comment (adelaidaa) [inline src/main/kotlin/uk/gov/justice/digital/assessments/services/AssessmentUpdateService.kt:71]: we could add the predictors as a parameter of this from builder and avoid mutating the updatedEpisode on line 77

File: src/main/kotlin/uk/gov/justice/digital/assessments/services/PredictorService.kt [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    val log: Logger = LoggerFactory.getLogger(this::class.java)
+  }
+
+  fun getPredictorResults(
+    assessmentSchemaCode: AssessmentSchemaCode,
+    episodeAnswers: Map<UUID, AnswersDto>,
+  ): List<PredictorScoreDto> {
+    val predictors = assessmentSchemaService.getPredictorsForAssessment(assessmentSchemaCode)
+
+    log.info("Found ${predictors.size} predictors for assessment type $assessmentSchemaCode")
```
Comment (adelaidaa) [inline src/main/kotlin/uk/gov/justice/digital/assessments/services/PredictorService.kt:27]: we could add the episode uuid here so we know which episode we are getting predictors for

File: src/test/kotlin/uk/gov/justice/digital/assessments/services/AssessmentUpdateServiceOASysTest.kt [hunk 3]
```diff
@@ -356,6 +358,7 @@ class AssessmentUpdateServiceOASysTest {
       )
     )
     every { assessmentUpdateRestClient.updateAssessment(any(), any(), any(), any()) } returns oasysError
+    every { predictorService.getPredictorResults(any(), any()) } returns emptyList()
```
Comment (adelaidaa) [inline src/test/kotlin/uk/gov/justice/digital/assessments/services/AssessmentUpdateServiceOASysTest.kt:361]: if possible i would avoid using any() for mocks as tests are more robust if we specify the right object inputs when mocking

**Review by adelaidaa — DISMISSED** [review]
Looks good!

File: src/main/kotlin/uk/gov/justice/digital/assessments/services/PredictorService.kt (hunk 2 above)
Comment (LucasCairns) [reply src/main/kotlin/uk/gov/justice/digital/assessments/services/PredictorService.kt]: 💯 - Added the Episode UUID to the log line

File: src/test/kotlin/uk/gov/justice/digital/assessments/services/AssessmentUpdateServiceOASysTest.kt (hunk 3 above)
Comment (LucasCairns) [reply src/test/kotlin/uk/gov/justice/digital/assessments/services/AssessmentUpdateServiceOASysTest.kt]: Tidied up this test 👍

File: src/main/kotlin/uk/gov/justice/digital/assessments/services/AssessmentUpdateService.kt (hunk 1 above)
Comment (LucasCairns) [reply src/main/kotlin/uk/gov/justice/digital/assessments/services/AssessmentUpdateService.kt]: Had to move a few things around but I think I've managed it 😄

File: src/test/kotlin/uk/gov/justice/digital/assessments/services/AssessmentUpdateServiceTest.kt [hunk 4]
```diff
@@ -43,15 +44,16 @@ class AssessmentUpdateServiceTest {
     courtCaseRestClient,
     assessmentUpdateRestClient,
     offenderService,
-    assessmentSchemaService
+    assessmentSchemaService,
   )
   private val assessmentUpdateService = AssessmentUpdateService(
```
Comment (SWhitlam) [inline src/test/kotlin/uk/gov/justice/digital/assessments/services/AssessmentUpdateServiceTest.kt:49]: This is getting a bit too large, we should look at refactoring it to make it more concise

**Review by SWhitlam — APPROVED** [review]

File: src/test/kotlin/uk/gov/justice/digital/assessments/services/AssessmentUpdateServiceTest.kt (hunk 4 above)
Comment (LucasCairns) [reply src/test/kotlin/uk/gov/justice/digital/assessments/services/AssessmentUpdateServiceTest.kt:49]: 💯  - there's quite a few dependencies being injected to this service, this comment in the tests says it all... `// Christ, what a lot of set up` 😅

**Pull request merged**